### PR TITLE
[klogs] Adjust klogs Plugin

### PR DESCRIPTION
**Pull request by staffbase-robert**
- Backend code is close to old implementation, with the following changes:
  - `pkg/klogs/parser` includes the code for the query parser. The query parser uses [`participle`](https://github.com/alecthomas/participle) for transforming the klogs query language into sql.
  - increased test coverage in the go code to >80%. Had to introduce an interface for the sql related code (to be able to mock `db.QueryContext` and `sql.Rows`.
  - implemented `MountCluster` and `MountHub` (which uses the common `pluginproxy` package for sending requests to the appropriate cluster satellite. 
- Frontend components are re-implemented with mui components. The following features are included:
  - syntax highlighting in the query input (powered by monaco editor)
  - query input supports autocomplete. columns and ops are suggested automatically
  -  improved the switching behavior between logs table page and aggregation page (search parameters are persisted and you don't have to open 2 tabs at the same time anymore)
  - improved aggregation page. introduced error message, which tells the user about required fields
  - every change in the aggregation page re-renders the chart. No need to click on a submit button.
  - charts have been implemented with victory. 
  - extracted the tooltip component to `packages/core`. Both plugin prometheus and klogs use the same tooltip component individual tooltip content components.
  - test coverage >80%

- [ ] I adjusted the corresponding [documentation](https://github.com/kobsio/kobs/tree/main/docs) for this change.
- [ ] I adjusted the [values.yaml for the hub](https://github.com/kobsio/kobs/blob/main/deploy/helm/hub/values.yaml) / [values.yaml for the satellite](https://github.com/kobsio/kobs/blob/main/deploy/helm/satellite/values.yaml) file and the [documentation for all values](https://github.com/kobsio/kobs/blob/main/docs/getting-started/installation/helm.md).

### CODE REVIEW

File: app/packages/klogs/src/components/page/LogsPage.tsx [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        search.fields[from],
+        ...search.fields.slice(to + 1),
+      ],
+    });
+  };
+
+  return (
+    <Page
+      title="klogs"
+      description="Fast, scalable and reliable logging using Fluent Bit and ClickHouse."
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsPage.tsx]: The description can also be confiured by the user. Please have a look how this is handled by the other plugins

File: app/packages/klogs/src/components/page/LogsPage.tsx [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+      subtitle={instance.cluster}
+      toolbar={
+        <LogsToolbar
+          {...search}
+          handlers={{ onChangeTime: handleChangeTime, onSearch: handleSearch }}
+          instance={instance}
+        />
+      }
+      actions={
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsPage.tsx]: Please use the same styling as the other plugins are using

File: app/packages/klogs/src/components/page/LogsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    >
+      <UseQueryWrapper
+        {...queryResult}
+        errorTitle="Failed to load applications"
+        isNoData={!queryResult.data || queryResult.data.documents === null}
+        noDataTitle="No logs found"
+        noDataMessage="There were no logs found for your search query"
+      >
+        <Stack direction="row" spacing={2} sx={{ maxWidth: '100%' }}>
+          <Paper>
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsPage.tsx]: Please replace `Paper` with `Card` to remove this ugly bottom shadow

File: app/packages/klogs/src/components/page/LogsPage.tsx [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+          </Paper>
+          <Stack direction="column" sx={{ width: '100%' }} spacing={2}>
+            <Paper>
+              <Stack direction="row" justifyContent="space-between" m={2}>
+                <Typography fontWeight="medium">{`${queryResult.data?.count} documents in ${queryResult.data?.took} milliseconds`}</Typography>
+                {queryResult.data?.documents && (
+                  <LogsDownload rows={queryResult.data.documents} fields={search.fields} fileDownload={fileDownload} />
+                )}
+              </Stack>
+              <Box sx={{ height: '150px' }}>
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsPage.tsx]: I would use the same size as in the existing klogs implementation: `250px`

File: app/packages/klogs/src/components/page/LogsPage.tsx [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            <IconButton component="a" href="https://kobs.io/main/plugins/klogs/" target="_blank">
+              <Description />
+            </IconButton>
+          </Tooltip>
+        </>
+      }
+    >
+      <UseQueryWrapper
+        {...queryResult}
+        errorTitle="Failed to load applications"
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsPage.tsx]: I think this is not about applications^^

File: app/packages/klogs/src/components/page/LogsPage.tsx [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+          <Tooltip title="Documentation">
+            <IconButton component="a" href="https://kobs.io/main/plugins/klogs/" target="_blank">
+              <Description />
+            </IconButton>
+          </Tooltip>
+        </>
+      }
+    >
+      <UseQueryWrapper
+        {...queryResult}
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsPage.tsx]: Can we please use the same coding style across all plugins

File: app/packages/klogs/src/components/page/LogsBucketChart.tsx [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        padding={{ bottom: 25, left: 20, right: 20, top: 0 }}
+        scale={{ x: 'time', y: 'linear' }}
+        theme={theme}
+        width={chartSize.width}
+        height={chartSize.height}
+        domain={getDomain(x, y)}
+      >
+        <VictoryAxis dependentAxis={false} />
+        <VictoryBar data={data} name="count" />
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsBucketChart.tsx]: Please reuse the properties from the existing implementation to remove these ugly gaps between the bars.

File: app/packages/klogs/src/components/page/LogsBucketChart.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  buckets: IBucket[];
+  onChangeTimeframe: (payload: IChangeTimeframePayload) => void;
+}
+
+const BrushVoronoiContainer = createContainer<VictoryVoronoiContainerProps, VictoryBrushContainerProps>(
+  'voronoi',
+  'brush',
+);
+
+// utility for creating a bar chart theme from the base Theme definition
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsBucketChart.tsx]: Please use the same coding style as in the other plugins for comments.

File: app/packages/klogs/src/components/page/LogsBucketChart.tsx [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+          />
+        }
+        padding={{ bottom: 25, left: 20, right: 20, top: 0 }}
+        scale={{ x: 'time', y: 'linear' }}
+        theme={theme}
+        width={chartSize.width}
+        height={chartSize.height}
+        domain={getDomain(x, y)}
+      >
+        <VictoryAxis dependentAxis={false} />
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsBucketChart.tsx]: There is a helper function to format the time axis please use it

File: app/packages/klogs/src/components/page/InternalEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import { MUIEditor } from '@kobsio/core';
+import { InputBaseComponentProps } from '@mui/material';
+import { forwardRef } from 'react';
+
+/**
+ * The `InternalEditor` component is a wrapper around our `MUIEditor` component, which allows us to use the editor
+ * within a `TextField` component of MUI.
+ */
+const InternalEditor = forwardRef<HTMLInputElement, InputBaseComponentProps>(function InternalEditor(props, ref) {
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/InternalEditor.tsx]: Can we name this just `Editor` like in the other plugins

File: app/packages/core/src/components/utils/editor/Editor.tsx [hunk 11]
```diff
@@ -40,6 +40,9 @@ export const Editor: FunctionComponent<IEditorProps> = ({ language, readOnly = f
         fontFamily: 'monospace',
         readOnly: readOnly,
         scrollBeyondLastLine: false,
+        scrollbar: {
```
Comment (ricoberger) [inline app/packages/core/src/components/utils/editor/Editor.tsx:59]: Why is this required?

File: app/packages/core/src/components/utils/editor/Editor.tsx [hunk 12]
```diff
@@ -87,6 +90,10 @@ export const MUIEditor: FunctionComponent<IMUIEditorProps> = ({
       default:
         break;
     }
+
+    if (language === 'klogs') {
```
Comment (ricoberger) [inline app/packages/core/src/components/utils/editor/Editor.tsx]: There is a switch case statement three lines above to handle the language setup

File: app/packages/core/src/components/utils/editor/klogs.ts [hunk 13]
```diff
@@ -0,0 +1,178 @@
+/* eslint-disable sort-keys */
+import { languages } from 'monaco-editor';
```
Comment (ricoberger) [inline app/packages/core/src/components/utils/editor/klogs.ts]: There is a `languages` folder for these files.

Please be not lazy and fix the eslint stuff instead of disabling it.

File: app/packages/klogs/src/components/page/LogsToolbar.tsx [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+      },
+    }),
+  );
+
+  useEffect(() => {
+    setQuery(initialQuery);
+  }, [initialQuery]);
+
+  return (
+    <Stack direction="row" component="form" onSubmit={handleSubmit} spacing={4} alignItems="center">
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsToolbar.tsx]: Please use the `Toolbar` component.

File: app/packages/klogs/src/components/page/LogsToolbar.tsx [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+      <TextField
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsToolbar.tsx]: I'm not sure what you did here, but somewho the TextField seems to use the nord theme instead of the material ui theme

File: app/packages/core/src/components/utils/Options.tsx [hunk 16]
```diff
@@ -201,16 +202,20 @@ export const Options: FunctionComponent<IOptionsProps> = ({
         <>
           {showSearchButton ? (
             <Button variant="contained" color="primary" onClick={() => setShow(true)}>
-              {times.time === 'custom'
-                ? `${formatTimestamp(times.timeStart)} to ${formatTimestamp(times.timeEnd)}`
-                : timeOptions[times.time].label}
+              <Typography noWrap={true}>
```
Comment (ricoberger) [inline app/packages/core/src/components/utils/Options.tsx]: This shouldn't be required

File: app/packages/core/src/components/utils/Options.tsx [hunk 17]
```diff
@@ -201,16 +202,20 @@ export const Options: FunctionComponent<IOptionsProps> = ({
...
+                  : timeOptions[times.time].label}
+              </Typography>
             </Button>
           ) : (
             <ButtonGroup variant="contained" color="primary">
               <Button onClick={() => setShow(true)}>
-                {times.time === 'custom'
-                  ? `${formatTimestamp(times.timeStart)} to ${formatTimestamp(times.timeEnd)}`
-                  : timeOptions[times.time].label}
+                <Typography noWrap={true}>
```
Comment (ricoberger) [inline app/packages/core/src/components/utils/Options.tsx]: Same this shouldn't be required

File: app/packages/core/src/utils/times.ts [hunk 18]
```diff
@@ -52,17 +52,20 @@ export const times = [
   'last90Days',
 ];
 
+export type TTimeQuick = Exclude<TTime, 'custom'>;
```
Comment (ricoberger) [inline app/packages/core/src/utils/times.ts]: I wouldn't do this, if you think this is useful please create another PR

File: app/packages/core/src/components/utils/Link.tsx [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import { forwardRef } from 'react';
+import { Link as RouterLink, LinkProps as RouterLinkProps } from 'react-router-dom';
+
+/**
+ * `Link` is a helper function, which we need to use the `Link` (imported as `RouterLink`) component from React Router
+ * within the `ListItem` component from Material UI.
+ */
+export const Link = forwardRef<HTMLAnchorElement, RouterLinkProps>(function link(itemProps, ref) {
```
Comment (ricoberger) [inline app/packages/core/src/components/utils/Link.tsx]:
```suggestion
export const Link = forwardRef<HTMLAnchorElement, RouterLinkProps>(function Link(itemProps, ref) {
```

File: app/packages/core/src/components/dashboards/Dashboards.tsx [hunk 20]
```diff
@@ -53,7 +53,7 @@ interface IDashboardGridProps {
  * The `DashboardGrid` component is used to render a list of `panels` in a responsive grid layout. We are using the
  * `react-grid-layout` package to render the grid, so that a user can also drage and resize the panels in the app.
  */
-const DashboardGrid: FunctionComponent<IDashboardGridProps> = ({ panels, times, setTimes }) => {
+const DashboardGrid: FunctionComponent<IDashboardGridProps> = ({ panels, setTimes, times }) => {
```
Comment (ricoberger) [inline app/packages/core/src/components/dashboards/Dashboards.tsx:56]: Why did you change this, how is this related to your PR?

File: app/packages/core/src/utils/hooks/useQueryState.tsx [hunk 21]
```diff
@@ -52,12 +57,7 @@ export const useQueryState = <S extends QueryState = QueryState>(initialState?:
     // parameter contains a valid time, so that we are really using the "last 15 minutes" and not the times from the
     // `timeEnd` and `timeStart` parameters.
     if (initialQueryFromUrl.current === true) {
-      if (
-        'time' in parsedQueryString &&
-        typeof parsedQueryString.time === 'string' &&
-        times.includes(parsedQueryString.time) &&
-        parsedQueryString.time !== 'custom'
-      ) {
+      if (queryStringHasTime(parsedQueryString)) {
```
Comment (ricoberger) [inline app/packages/core/src/utils/hooks/useQueryState.tsx]: Where is this function used again, that we can not leave the statement inline?

File: app/packages/core/src/crds/dashboard.ts [hunk 22]
```diff
@@ -1,6 +1,8 @@
+import { TTimeQuick } from '../utils/times';
+
 export interface IDashboard {
   cluster: string;
-  defaultTime?: string;
+  defaultTime?: TTimeQuick;
```
Comment (ricoberger) [inline app/packages/core/src/crds/dashboard.ts]: Same as in your other PR, I wouldn't do this here.

File: app/packages/klogs/src/components/common/sortFields.ts [hunk 23]
```diff
@@ -0,0 +1,30 @@
+// TODO: make this configurable in plugin options
+/* eslint-disable sort-keys */
+const columnPriority: Partial<Record<string, number>> = {
```
Comment (ricoberger) [inline app/packages/klogs/src/components/common/sortFields.ts]: I wouldn't add this feature, since the default fields do not make much sense to display them on the top, since most of the time we already filter for these fields. Instead I would add the Textfield to filter the list of fields like discussed last time.

File: app/packages/klogs/src/components/page/LogsTable.tsx [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+          <TableRow>
+            <TableCell />
+            {fields.map((field) => (
+              <TableCell key={field}>
+                <TableSortLabel
+                  onClick={() => handlers.onChangeSort(field)}
+                  active={orderBy === field}
+                  direction={orderBy === field ? order : 'asc'}
+                >
+                  <Typography>{field}</Typography>
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsTable.tsx]: Please align this on the top

File: app/packages/klogs/src/components/page/LogsDownload.tsx [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+        csv = csv + ';' + (row.hasOwnProperty(field) ? row[field] : '-');
+      }
+
+      csv = csv + '\r\n';
+    }
+
+    fileDownload(csv, 'kobs-export-logs.csv');
+  };
+
+  return (
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsDownload.tsx:51]: Can we please use the same styling for actions as in the other plugins via the Menu component

File: app/packages/klogs/src/components/page/LogsRowPreview.tsx [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * RowPreview renders a preview of a log document with pre-defined fields
+ */
+const RowPreview: FunctionComponent<IRowPreviewProps> = ({ row }) => {
+  const knownColumns = ['cluster', 'namespace', 'app', 'pod_name', 'container_name', 'host'];
+  const contentColumns = Object.keys(row).filter((col) => col.startsWith('content_'));
+
+  return (
+    <TableCell size="small">
+      <Typography
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsRowPreview.tsx:21]: I wouldn't cap this to two lines and instead use the existing logic to display some more information

File: app/packages/klogs/src/components/page/LogsToolbar.tsx [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+          inputProps: {
+            callSubmit: () => {
+              handlers.onSearch(query);
+            },
+            loadCompletionItems: loadCompletions,
+          },
+        }}
+        fullWidth={true}
+      />
+      <Options
```
Comment (ricoberger) [inline app/packages/klogs/src/components/page/LogsToolbar.tsx]: The existing additional options are missing here

File: app/packages/core/src/components/utils/editor/Editor.tsx (hunk 11 above)
Comment (staffbase-robert) [reply app/packages/core/src/components/utils/editor/Editor.tsx:59]: The editor would otherwise "consume the scrollwheel". The user would have to move the mouse outside of the editor window to be able to resume scrolling the page. The user should never have to scroll inside the editor, because we set the height to match the line count. 
I think the change should be harmless for the other places where we use this editor. If not, I can make this an optional prop.

File: app/packages/klogs/src/components/page/LogsPage.tsx (hunk 2 above)
Comment (staffbase-robert) [reply app/packages/klogs/src/components/page/LogsPage.tsx:161]: The dropdown requires an additional click. IMO the user experience is better, when there are fewer steps involved. Have you tried it?

File: app/packages/klogs/src/components/page/LogsPage.tsx (hunk 6 above)
Comment (staffbase-robert) [reply app/packages/klogs/src/components/page/LogsPage.tsx]: What's wrong with spread syntax for react components?

It will be difficult to keep up with your preferred coding style, if we don't add linting rules for things like this.

File: app/packages/core/src/components/utils/editor/klogs.ts (hunk 13 above)
Comment (staffbase-robert) [reply app/packages/core/src/components/utils/editor/klogs.ts]: This has nothing to do with laziness, but with the readability of the file.
Sometimes I want a property called `closed` to be after the property `open`, because it makes more sense compared to the lexical ordering. I will reformat the file, because I don't have the energy to argue this.

File: app/packages/core/src/components/dashboards/Dashboards.tsx (hunk 20 above)
Comment (staffbase-robert) [reply app/packages/core/src/components/dashboards/Dashboards.tsx:56]: We added this at the same time, after a rebase I resolved the merge conflicts and preferred the lexical order.

File: app/packages/core/src/utils/hooks/useQueryState.tsx (hunk 21 above)
Comment (staffbase-robert) [reply app/packages/core/src/utils/hooks/useQueryState.tsx]: This function isn't about re-usablity, but about it being a [`type-guard`](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards).

I improved the types inside `times.ts` by narrowing down the type of `timeOptions`. Instead of it being a `Recod<string, ...>`, I made it a `Record<TTimeQuick, ...>`. This is more precise and should be preferred. 

The `Record<string,...>` version wouldn't give us type-errors when timeOptions is accessed with a string like so:
```ts
// type of option is { label: string; seconds: number; }, but at runtime it's actually undefined
const option = timeOptions["unknownProperty"]
```

File: app/packages/klogs/src/components/page/LogsTable.tsx (hunk 24 above)
Comment (staffbase-robert) [reply app/packages/klogs/src/components/page/LogsTable.tsx]: Could you explain this change request? I don't understand what you mean by "on the top".

**Review by ricoberger — APPROVED** [review]
